raiden/routing.py

```diff
@@ -1,29 +1,123 @@
 from heapq import heappop, heappush
-from typing import List
+from typing import Any, Dict, List, Tuple
 
 import networkx
+import requests
 import structlog
+from eth_utils import to_canonical_address, to_checksum_address
 
+from raiden.constants import DEFAULT_HTTP_REQUEST_TIMEOUT
 from raiden.transfer import channel, views
 from raiden.transfer.state import (
     CHANNEL_STATE_OPENED,
     NODE_NETWORK_REACHABLE,
     NODE_NETWORK_UNKNOWN,
     ChainState,
+    NettingChannelState,
     RouteState,
 )
 from raiden.utils import pex, typing
 
 log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
 
 
+def check_channel_constraints(
+        channel_state: NettingChannelState,
+        from_address: typing.InitiatorAddress,
+        partner_address: typing.Address,
+        amount: int,
+        network_statuses: Dict[typing.Address, str],
+        routing_module: str,
+) -> bool:
+    # check channel state
+    if channel.get_status(channel_state) != CHANNEL_STATE_OPENED:
+        log.info(
+            'Channel is not opened, ignoring',
+            from_address=pex(from_address),
+            partner_address=pex(partner_address),
+            routing_source=routing_module,
+        )
+        return False
+
+    # check channel distributable
+    distributable = channel.get_distributable(
+        channel_state.our_state,
+        channel_state.partner_state,
+    )
+
+    if amount > distributable:
```

Comment:
The checks for distributable amount and network status are transient, meaning both values can change and become valid. The route should stay as a possible route, and the state machine should filter out the ones that are not usable:

https://github.com/raiden-network/raiden/issues/2817

Important for mediators. If the mediator receives a mediated transfer but all the routes forward are unavailable the transfer will get stuck, to unstuck it the routes must be available to the state machine, so that it can continue once either of these attributes change.

https://github.com/raiden-network/raiden/issues/1675

To update the state machine

Reply:
BTW, I'm not saying the above must be done in this PR, I'm just bringing it up.

Reply:
Yes, good points. This is one of the reasons why I refactored this stuff out, so we only have to change it in one place.

I think this should be done in different PRs though.